### DISCUSSION TITLE
Stop liquidsoap and icecast2 before infinite loop

### DISCUSSION THREAD

**Discussion by juanamm** [savonet/liquidsoap#3671] [User Support]
Hi there,

I need to know if there is any script or way to stop the server automatically in the event of an infinite loop error.

I use liquidsoap as a daemon (Thanks to @vitoyucepi's help [here] )(https://github.com/savonet/liquidsoap/discussions/3354#discussioncomment-6880550)

Icecast version: Icecast 2.4.4
Liquidsoap version: Liquidsoap 2.1.4 (installed with OCaml Package Manager (opam))

icecast2@ip-17x-xx-xx-xxx:~/liquidsoap-daemon$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.3 LTS
Release:        22.04
Codename:       jammy
Grateful for the help you can give me.

**Comment by vitoyucepi** [comment]
Hi @juanamm,
Can you please explain what kind of infinite loop you're experiencing?

**Comment by juanamm** [comment]
Now it's too late for me and I'm in bed, but I mean any liquidsoap
error, they all cause that. A few days ago I had an error because I
had not renewed or SSL certificate, the server's CPU was at 100%. I
want to avoid that and if an error causes the streaming to not work I
prefer that icecast and liquidsoap be turned off until I detect it and
not that the server punishes the CPU.

**Comment by vitoyucepi** [comment]
> and I'm in bed

Take a break, the asynchronous nature of the issues allows you to not answer questions right away.

> A few days ago I had an error because I had not renewed or SSL certificate, the server's CPU was at 100%.

I think this might be another problem, you should tell me more about it.

> I want to avoid that and if an error causes the streaming to not work I prefer that icecast and liquidsoap be turned off until I detect it and not that the server punishes the CPU.

This looks like a job for a sidecar service or something. Serious IT infrastructure relies on a monitoring and alerting combines like grafana, but it can be done on your own with a much simpler script.

**Comment by juanamm** [comment]
That's exactly what I'm looking for, a simple script that stops the icecast
and liquidsoap services in the event of any error.
Until now, any syntax error or something failing generates an infinite
loop, how do I know?
Because my server's CPU goes up to 100% and if I look at 'journalctl'
errors are still being generated every second.

Just for testing purposes can you replicate the error that happened when I
was on vacation, the SSL certificate referenced by the Icecast XML file
expired and the streaming stopped working, however the AWS CPU at 100%
alarms still occurred generating because the services continued to run with
the error.
All you have to do is change the path or name of the PEM file in the
Icecast XML so that the streaming that I have configured in HTTPS stops
working and generates the error.
This is the line I am referring to for the purposes of testing:
<ssl-certificate>/usr/share/icecast2/bundle.pem</ssl-certificate>
What I did was change the name "bundle.pem" to "bundle666.pem" so that the
error was generated.

The idea is that the script detects any error, not just the example I
shared before.

Thank you for any help you can give me.

P.S.
I forgot to share my liquidsoap script with you so you can see how I
configured SSL streaming:

# Settings
# settings.server.telnet := true
# settings.server.telnet.port := 8595
settings.log.level := 3
settings.harbor.bind_addrs := ["0.0.0.0"]
settings.log.file := true
settings.log.file.path := "/home/icecast2/liquidsoap/log/registro.log"


# SSL certificate configuration
transport = http.transport.tls(
  # Server mode: required,
  # client mode: optional, add certificate to trusted pool
  certificate="/etc/letsencrypt/live/osirisfm.org/fullchain.pem",
  # Server mode: required, client mode: ignored
  key="/etc/letsencrypt/live/osirisfm.org/privkey.pem",
)

# Environment Variables
password_dj=environment.get("PASSWORD_DJ")
password_inv1=environment.get("PASSWORD_INV1")
password_inv2=environment.get("PASSWORD_INV2")
password_debby=environment.get("PASSWORD_DEBBY")
source_passw=environment.get("SOURCE_PASSW")
port_https=environment.get("PORT_HTTPS")
port_live=environment.get("PORT_LIVE")
port_inv1=environment.get("PORT_INV1")
port_inv2=environment.get("PORT_INV2")
port_debby=environment.get("PORT_DEBBY")
host_https_no_prot=environment.get("HOST_HTTPS_NO_PROT")

# Convert string to integer
port_live_int=int_of_string(port_live)
port_https_int=int_of_string(port_https)
port_inv1_int=int_of_string(port_inv1)
port_inv2_int=int_of_string(port_inv2)
port_debby_int=int_of_string(port_debby)


# Registrar el valor en el archivo de registo
log("Valor de port_live: " ^ port_live)


# Definir la fuente original de la lista de reproducción y aplicar nrj
original_playlist_source=playlist("~/music/playlist.pls")
original_playlist=nrj(original_playlist_source)

# Aplicar crossfades a la playlist
original_playlist = crossfade(fade_out=3.0, fade_in=3.0, original_playlist)

# Security fallback
security_source=single("~/music_security/security.mp3")
playlist_securitized=fallback(track_sensitive=false, [original_playlist,
security_source])

# Support for live streams
live=input.harbor(
    "live",
    port=port_live_int,
    password=password_dj,
    replay_metadata=true,
    icy_metadata_charset="UTF-8"
)
inv1=input.harbor(
    "inv1",
    port=port_inv1_int,
    password=password_inv1,
    replay_metadata=true,
    icy_metadata_charset="UTF-8"
)
inv2=input.harbor(
    "inv2",
    port=port_inv2_int,
    password=password_inv2,
    replay_metadata=true,
    icy_metadata_charset="UTF-8"
)
debby=input.harbor(
    "debby",
    port=port_debby_int,
    password=password_debby,
    replay_metadata=true,
    icy_metadata_charset="UTF-8"
)

# Fallback for live and music playlists
radio=fallback(track_sensitive=false, [live, inv1, inv2, debby,
playlist_securitized])

# Stream it out MP3
output.icecast(
    %ffmpeg(format="mp3", %audio(codec="libmp3lame", b="192k")),
    radio,
    mount="stream",
    name="DJ OsiriS Music Service",
    description="DJ OsiriS FM, estamos las 24 horas online con una variada
propuesta musical y DJs en vivo!",
genre="El que tú sueñes :)",
    fallible=true,
    host=host_https_no_prot,
    port=port_https_int,
    password=source_passw,
    transport=transport,
)

*NOTE: *
I am providing this to you for the sole purpose of allowing you to
replicate the error for testing purposes. If this is not necessary for you
and you have a snippet of some code that detects errors generically in
icecast and liquidsoap, share it with me and I will test it.


Thanks again :)

**Reply by vitoyucepi** [reply]
I just noticed that you're using version *2.2.x*, not *2.1.4* as you described, because of `http.transport.tls`.

Let me describe my experience with certificates.
1. If the certificate expires during use, liquidsoap will not complain about it.
   On reconnect it will show a `TLS handshake error: certificate expired`.
   After that it will retry the connection in 3 seconds.
1. If certificate is self-signed, liquidsoap will print an error `TLS handshake error: bad certificate`.
   You can fix this by defining a custom certificate in transport:
   ```ruby
   transport = http.transport.tls(
     certificate="cert.pem",
   )
   ```
1. The only way to get liquidsoap to use 100% of the CPU is to use encryption when trying to connect to an unencrypted socket.

In the first and second cases the CPU usage is low, there's no infinite loop or anything like that.
In the third case, the main thread is stuck in an infinite loop.

@juanamm,
If you don't want to rely on LE certificates, I can recommend a few methods that will prevent their use in this case.
1. If your traffic doesn't go through the public networks, you can use an unencrypted connection to icecast.
1. If you have multiple servers, you can set up a VPN between them, e.g. wireguard, and still use an unencrypted connection.
1. Create a self-signed certificate with an expiration date of 10 years.

**Comment by toots** [comment]
I'm moving this to a discussion. Monitoring is great and should be required for any production system but we should also not go into infinite loop. If you've got something to reproduce I'd be happy to look at fixing it in the `rolling-release-v2.2.x` branch.

**Reply by juanamm** [reply]
I think that the previous daemon that was discontinued was somewhat more solid, if an error occurred, be it syntax or any other nature, the service would stop.
The daemon that @vitoyucepi  helped me configure is excellent, but in the event of errors the service continues to run and the infinite loop is formed because the script reaches the error and starts from the beginning, until it reaches the error again and starts again.
I have verified this because the CPU reaches 100% on a basic AWS machine (t2.micro) and also by checking the journalctl logs where the error is shown every second.

**Reply by vitoyucepi** [reply]
@juanamm,
Could you please post the journalctl logs?

**Comment by juanamm** [comment]
> @juanamm, Could you please post the journalctl logs?

My journalctl are now clean, I have no errors.
It should "make" a syntax error or like the one I told you about the SSL certificate to be able to show you errors.....

**Comment by juanamm** [comment]
Conclusion: I need a daemon or a script that stops the services in the event of any error in the icecast or liquidsoap services.

**Reply by vitoyucepi** [reply]
Ok, here's my idea. In case of an infinite loop, liquidsoap will certainly fail to respond to the http API request.

```ruby
def register_server_healthcheck(port=8000)
  def handler(_)
    http.response(
      status_code=200,
      content_type="application/json; charset=UTF-8",
      data='{"status":"success","data":{"date":#{int(time())}}}\n',
    )
  end

  harbor.http.register.simple('/healthcheck', handler, method="GET", port=port)
end

register_server_healthcheck()
```

You should write a script that makes an HTTP request to something like `http://liquidsoap:8000/healthcheck`.

**Comment by juanamm** [comment]
> Ok, here's my idea. In case of an infinite loop, liquidsoap will certainly fail to respond to the http API request.
> 
> ```ruby
> def register_server_healthcheck(port=8000)
>   def handler(_)
>     http.response(
>       status_code=200,
>       content_type="application/json; charset=UTF-8",
>       data='{"status":"success","data":{"date":#{int(time())}}}\n',
>     )
>   end
> 
>   harbor.http.register.simple('/healthcheck', handler, method="GET", port=port)
> end
> 
> register_server_healthcheck()
> ```
> 
> You should write a script that makes an HTTP request to something like `http://liquidsoap:8000/healthcheck`.


First of all, thank you for trying, but unfortunately it doesn't work, could not detect an error generated for testing purposes when modifying the Icecast2 XML on the line of the name of the file that contains the SSL certificates


Correct Icecast.XML LINE:
`<ssl-certificate>/usr/share/icecast2/bundle.pem</ssl-certificate>`

Log before creating the error (/var/log/icecast2/error.log):

```
[2024-01-30  16:41:46] INFO main/main Icecast 2.4.4 server started
[2024-01-30  16:41:46] INFO yp/yp_recheck_config Adding new YP server "http://dir.xiph.org/cgi-bin/yp-cgi" (timeout 6s, default interval 30s)
[2024-01-30  16:41:46] INFO yp/yp_update_thread YP update thread started
[2024-01-30  16:41:46] INFO connection/get_ssl_certificate SSL certificate found at /usr/share/icecast2/bundle.pem
[2024-01-30  16:41:46] INFO connection/get_ssl_certificate SSL using ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-RSA-DES-CBC3-SHA:ECDHE-ECDSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:DES-CBC3-SHA:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA
[2024-01-30  16:41:57] INFO connection/_handle_source_request Source logging in at mountpoint "/stream" from 18.231.22.106
[2024-01-30  16:41:57] WARN format/format_get_type Unsupported or legacy stream type: "audio/mpeg". Falling back to generic minimal handler for best effort.
[2024-01-30  16:41:57] INFO admin/admin_handle_request Received admin command metadata on mount "/stream"
[2024-01-30  16:41:57] INFO util/util_conv_string converting metadata from UTF-8 to UTF-8
[2024-01-30  16:41:57] INFO admin/command_metadata Metadata on mountpoint /stream changed to "Hillsong 日本語 - ディス・イズ・リビング"
[2024-01-30  16:41:57] INFO source/source_main listener count on /stream now 0
[2024-01-30  16:42:09] INFO source/source_main listener count on /stream now 1
```


Healtcheck result without error: {"status":"success","data":{"date":1706643860}}


HTOP screenshot (normal CPU consumption)
![CPU_normal](https://github.com/savonet/liquidsoap/assets/48572613/34215a53-7800-4725-957a-28cf8237e782)


Incorrect Icecast.XML LINE to cause error:
`<ssl-certificate>/usr/share/icecast2/bundle666.pem</ssl-certificate>`


Log after creating the error with the SSL certificat (modifying the Icecast XML)

```
[2024-01-30  16:47:15] INFO main/main Icecast 2.4.4 server started
[2024-01-30  16:47:15] INFO yp/yp_recheck_config Adding new YP server "http://dir.xiph.org/cgi-bin/yp-cgi" (timeout 6s, default interval 30s)
[2024-01-30  16:47:15] WARN connection/get_ssl_certificate Invalid cert file /usr/share/icecast2/bundle666.pem
[2024-01-30  16:47:15] INFO connection/get_ssl_certificate No SSL capability on any configured ports
[2024-01-30  16:47:15] INFO stats/_stats_thread stats thread started
[2024-01-30  16:47:15] INFO yp/yp_update_thread YP update thread started
[2024-01-30  16:48:16] INFO main/_server_proc Caught halt request, shutting down...
[2024-01-30  16:48:16] INFO main/main Shutting down
[2024-01-30  16:48:16] INFO fserve/fserve_shutdown file serving stopped
[2024-01-30  16:48:17] INFO slave/_slave_thread shutting down current relays
[2024-01-30  16:48:17] INFO slave/_slave_thread Slave thread shutdown complete
[2024-01-30  16:48:17] INFO auth/auth_shutdown Auth shutdown
[2024-01-30  16:48:17] INFO yp/yp_shutdown YP thread down
[2024-01-30  16:48:17] INFO stats/stats_shutdown stats thread finished
```

Healtcheck result with error:  `{"status":"success","data":{"date":1706644186}}`

_Conclusion:_ It does not detect the error generated.

I also simulated a syntax error in main.liq by omitting the closing of a curved parenthesis ")" like this at the beginning of the script:

```
# SSL certificate configuration
transport = http.transport.tls(
  # Server mode: required,
  # client mode: optional, add certificate to trusted pool
  certificate="/etc/letsencrypt/live/osirisfm.org/fullchain.pem",
  # Server mode: required, client mode: ignored
  key="/etc/letsencrypt/live/osirisfm.org/privkey.pem",

.......
```

Journalctl LOG:
```
Jan 30 16:55:44 ip-000-31-13-xxx liquidsoap[3070]: At /home/icecast2/liquidsoap/main.liq, line 21, char 13:
Jan 30 16:55:44 ip-000-31-13-xxx liquidsoap[3070]: password_inv1=environment.get("PASSWORD_INV1")
Jan 30 16:55:44 ip-000-31-13-xxx liquidsoap[3070]: Error 2: Parse error
-----
```

HTOP screenshot (CPU 100%)

![CPU](https://github.com/savonet/liquidsoap/assets/48572613/9db38b4c-3c4f-4933-b7de-7daa32df9077)



Status Liquidsoap
 sudo systemctl status liquidsoap@main.service
● liquidsoap@main.service - Liquidsoap daemon for main
     Loaded: loaded (/etc/systemd/system/liquidsoap@.service; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system/liquidsoap@main.service.d
             └─override.conf
     Active: active (running) since Tue 2024-01-30 17:01:06 -03; 2s ago
       Docs: https://www.liquidsoap.info/
   Main PID: 3174 (liquidsoap)
      Tasks: 1 (limit: 1121)
     Memory: 143.9M
        CPU: 2.584s
     CGroup: /system.slice/system-liquidsoap.slice/liquidsoap@main.service
             └─3174 /home/icecast2/.opam/default/bin/liquidsoap /home/icecast2/liquidsoap/main.liq

Jan 30 17:01:06 ip-000-31-13-xxx systemd[1]: Started Liquidsoap daemon for main.

Status Icecast2
 sudo systemctl status icecast2
● icecast2.service - LSB: Icecast2 streaming media server
     Loaded: loaded (/etc/init.d/icecast2; generated)
     Active: active (running) since Tue 2024-01-30 16:54:06 -03; 8min ago
       Docs: man:systemd-sysv-generator(8)
    Process: 3051 ExecStart=/etc/init.d/icecast2 start (code=exited, status=0/SUCCESS)
      Tasks: 4 (limit: 1121)
     Memory: 3.1M
        CPU: 146ms
     CGroup: /system.slice/icecast2.service
             └─3059 /usr/bin/icecast2 -b -c /etc/icecast2/icecast.xml

Jan 30 16:54:06 ip-000-31-13-xxx systemd[1]: Starting LSB: Icecast2 streaming media server...
Jan 30 16:54:06 ip-000-31-13-xxx icecast2[3051]:  * Starting streaming media server icecast2
Jan 30 16:54:06 ip-000-31-13-xxx icecast2[3051]:    ...done.
Jan 30 16:54:06 ip-000-31-13-xxx systemd[1]: Started LSB: Icecast2 streaming media server.

Since the code for the register_server_healthcheck function is at the end of the script, it is not executed and I am left wondering whether or not it would detect the error:

Healtcheck Result: 
This website cannot be accessed. Page xx.231.xx.189 has refused the connection.



I don't know how to make this code work right, if you can figure it out I want to suggest some additional ideas that I think are missing.
This function should run automatically with liquidsoap and monitor at least once for any errors.
It would also be necessary to add a conditional in case the result is not successful and in this case stop the services.


I made a mistake in my initial post when sharing my Liquidsoap version's
```
liquidsoap --version
Liquidsoap 2.2.0+dev
Copyright (c) 2003-2023 Savonet team
Liquidsoap is open-source software, released under GNU General Public License.

```
PS
My apologies for my bad English and for the length of my answer.

**Reply by vitoyucepi** [reply]
Let me try again. By the way, the problem was fixed in #3681.
It seems that it's not possible to tell if the server is stalling or not by that measure.

You can get the status of icecast streams from the `/status-json.xsl` file.
If the desired mount point is down, check if liquidsoap.service is running and stop it.
Here's an example in shell, I didn't test it. Maybe it would be better to rewrite it using python.
```bash
if ! curl 'http://icecast:8000/status-json.xsl' | grep 'http://icecast:8000/stream'; then
  if systemctl is-active liquidsoap.service; then
    systemctl stop liquidsoap.service
  fi
fi
```

**Comment by juanamm** [comment]
Following your logic, I created this script in Python, I tested it and it worked. I share it with you for whoever wants to test it.
**EDITED:** 
_This code is not complete; I forgot the most important part, which is stopping the Icecast and Liquidsoap services. This omission would cause an infinite loop and the CPU to spike to 100%. In my last post, you can find the corrected code._

```
#!/usr/bin/env python3
import requests
import subprocess
import time
import logging

# Configure the logging system
logging.basicConfig(filename='healthcheck.log', level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s')

# Maximum restarts allowed for Icecast2 and Liquidsoap
MAX_RESTARTS_ICECAST2 = 3
MAX_RESTARTS_LIQUIDSOAP = 3

# URL of the Icecast status file
icecast_status_url = 'http://icecast:8000/status-json.xsl'

def restart_service(service_name):
    logging.info(f"Restarting service {service_name}...")
    subprocess.run(['sudo', 'systemctl', 'restart', service_name])

while True:
    # Check the status of the mount point in Icecast
    try:
        response = requests.get(icecast_status_url, timeout=5)
        response.raise_for_status()  # Raise exception for HTTP errors
        icecast_data = response.json()
        server_name = icecast_data.get('icestats', {}).get('source', {}).get('server_name', '')
        
        if not server_name:
            logging.error("Streaming seems inactive. Restarting services...")

            # Restart Icecast2
            restart_service('icecast2')

            # Wait for 10 seconds before restarting Liquidsoap
            time.sleep(10)

            # Restart Liquidsoap
            restart_service('liquidsoap@main.service')
            
        else:
            logging.debug("Streaming is active. Everything is working correctly.")
            
    except requests.exceptions.RequestException as e:
        logging.error(f"Couldn't connect to the Icecast server: {e}")

    # Wait for 30 seconds before the next check
    time.sleep(30)
```


This Python script serves as a health check monitor for an Icecast streaming server. It periodically queries the Icecast server's status using a JSON API endpoint. If the streaming appears inactive or encounters an error, the script takes corrective actions by restarting the Icecast2 service, followed by a brief delay, and then restarting the Liquidsoap service. The script logs relevant events and errors to a file, providing a detailed record of the server's health and any actions taken. The periodic nature of the checks ensures ongoing monitoring. Additionally, the number of restart attempts is limited, and once exhausted, the script gracefully stops, preventing continuous execution in case of persistent issues.

**Reply by vitoyucepi** [reply]
I found another way to check the health of `output.icecast`.

If you assign its value to a variable `f = output.icecast()` and then use `f.time()`, you'll get the amount of data sent through output in seconds. Add it to the healthcheck response and you'll be able to calculate the time difference between two runs. If the ratio of output time difference to real time difference is less than 1, then the output is stalling.

But that's just a theory.

**Comment by juanamm** [comment]
Good point to test, but for now, I'll stick with the initial idea that works correctly.
I just made some important modifications; I added the feature to record the last 50 lines of the journalctl, Liquidsoap, and Icecast logs in order to understand why the script failed in case of an error and to have all the logs unified.

Then, I added the stopping of the Icecast and Liquidsoap services if the streaming fails and is not resolved with the 3 restarts.

The complete script looks like this:
```python
#!/usr/bin/env python3
import requests
import subprocess
import time
import logging

# Configure the logging system
logging.basicConfig(filename='healthcheck.log', level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s')

# Maximum restarts allowed for Icecast2 and Liquidsoap
MAX_RESTARTS_ICECAST2 = 3
MAX_RESTARTS_LIQUIDSOAP = 3

# URL of the Icecast status file
icecast_status_url = 'http://icecast:8000/status-json.xsl'

def restart_service(service_name):
    logging.info(f"Restarting service {service_name}...")
    subprocess.run(['sudo', 'systemctl', 'restart', service_name])

while True:
    # Check the status of the mount point in Icecast
    try:
        response = requests.get(icecast_status_url, timeout=5)
        response.raise_for_status()  # Raise exception for HTTP errors
        icecast_data = response.json()
        server_name = icecast_data.get('icestats', {}).get('source', {}).get('server_name', '')
        
        if not server_name:
            logging.error("Streaming seems inactive. Restarting services...")

            # Get the last 50 lines of journalctl logs
            journalctl_output = subprocess.run(['sudo', 'journalctl', '--lines', '50'], capture_output=True, text=True)
            logging.error(f"Journalctl logs:\n{journalctl_output.stdout}")

            # Get the last 50 lines of Icecast logs
            icecast_logs_output = subprocess.run(['sudo', 'tail', '-n', '50', '/var/log/nginx/error.log'], capture_output=True, text=True)
            logging.error(f"Icecast logs:\n{icecast_logs_output.stdout}")

            # Get the last 50 lines of Liquidsoap logs
            liquidsoap_logs_output = subprocess.run(['sudo', 'tail', '-n', '50', '/home/icecast2/liquidsoap/log/registro.log'], capture_output=True, text=True)
            logging.error(f"Liquidsoap logs:\n{liquidsoap_logs_output.stdout}")

            # Restart Icecast2
            restart_service('icecast2')

            # Wait for 10 seconds before restarting Liquidsoap
            time.sleep(10)

            # Restart Liquidsoap
            restart_service('liquidsoap@main.service')
            
        else:
            logging.debug("Streaming is active. Everything is working fine.")
            
    except requests.exceptions.RequestException as e:
        logging.error(f"Failed to connect to the Icecast server: {e}")

    # Wait for 30 seconds before the next check
    time.sleep(30)

    # Check if the maximum number of restarts for Icecast2 has been exceeded
    restart_count_icecast2 = 1 + (logging.getLogger().getEffectiveLevel() // 10)  # Divide by 10 to get a value similar to the number of attempts

    if restart_count_icecast2 > MAX_RESTARTS_ICECAST2:
        logging.warning(f"Maximum restarts for Icecast2 ({MAX_RESTARTS_ICECAST2}) reached. Stopping services and terminating the script.")
        
        # Stop Icecast2
        subprocess.run(['sudo', 'systemctl', 'stop', 'icecast2'])

        # Stop Liquidsoap
        subprocess.run(['sudo', 'systemctl', 'stop', 'liquidsoap@main.service'])
        
        break  # Stop the script
